engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/MethodDefinitions.scala

```diff
@@ -127,11 +135,43 @@ case object MethodDefinitions extends IRPass {
     ir.copy(bindings = withStaticAliases)
   }
 
+  private def addTypeAscriptionToSelfArgument(
+    methodBody: Expression,
+    typ: Expression
+  ): Expression = methodBody match {
+    case lambda: Function.Lambda =>
+      val newArguments = lambda.arguments match {
+        case (selfArg @ DefinitionArgument.Specified(
+              Name.Self(_, _, _, _),
+              _,
+              _,
+              _,
+              _,
+              _,
+              _
+            )) :: rest =>
+          selfArg.copy(ascribedType = Some(typ)) :: rest
```

Comment:
I was expecting fix in the `IrToTruffle` rather than in the IR passes. Probably because I am more comfortable with the Truffle nodes than the IR...